rootfs/etc/nginx/lua/util/dns.lua

```diff
@@ -0,0 +1,53 @@
+local resolver = require("resty.dns.resolver")
+local lrucache = require "resty.lrucache"
+local configuration = require("configuration")
+
+local _M = {}
+local CACHE_SIZE = 10000
+
+local cache, err = lrucache.new(CACHE_SIZE)
+if not cache then
+  return error("failed to create the cache: " .. (err or "unknown"))
+end
+
+function _M.resolve(host)
+  local cached_address = cache:get(host)
+  if cached_address then
+    ngx.log(ngx.INFO, string.format("address %s for host %s was resolved from cache", cached_address, host))
+    return cached_address
+  end
+
+  local r, err = resolver:new{
+    nameservers = configuration.nameservers,
+    retrans = 5,
+    timeout = 2000,  -- 2 sec
+  }
+
+  if not r then
+    ngx.log(ngx.ERR, "failed to instantiate the resolver: " .. tostring(err))
+    return host
+  end
+
+  local answers, err, _tries = r:query(host, { qtype = r.TYPE_A }, {})
+  if not answers then
+    ngx.log(ngx.ERR, "failed to query the DNS server: " .. tostring(err))
+    return host
+  end
+
+  if answers.errcode then
+    ngx.log(ngx.ERR, string.format("server returned error code: %s: %s", answers.errcode, answers.errstr))
+    return host
+  end
+
+  for _, ans in ipairs(answers) do
+    if ans.address then
+      cache:set(host, ans.address, ans.ttl)
+      return ans.address
```

Comment:
I wonder if we should use all A records here when there are more than one. Can not think of a reason why we would not wanna do that.